### PR TITLE
configure authorization service via yaml file

### PR DESCRIPTION
**Pull request by jkralik**

### CODE REVIEW

File: authorization/oauth/oauth.go [hunk 1]
```diff
@@ -30,22 +30,22 @@ func (a AuthStyle) ToOAuth2() oauth2.AuthStyle {
...
-	AuthURL   string    `json:"AuthUrl" envconfig:"AUTH_URL" env:"AUTH_URL"`
-	TokenURL  string    `json:"TokenUrl" envconfig:"TOKEN_URL" env:"TOKEN_URL"`
-	AuthStyle AuthStyle `envconfig:"AUTH_STYLE" env:"AUTH_STYLE"`
+type Config struct {
+	ClientID     string    `yaml:"clientID" json:"clientID"`
+	ClientSecret string    `yaml:"clientSecret" json:"clientSecret"`
+	Scopes       []string  `yaml:"scopes" json:"scopes"`
+	AuthURL      string    `yaml:"authorizationURL" json:"authorizationURL"`
+	TokenURL     string    `yaml:"tokenURL" json:"tokenURL"`
+	AuthStyle    AuthStyle `yaml:"authStyle" json:"authStyle"`
```
Comment (ondrejtomcik) [inline authorization/oauth/oauth.go:40]: there is no auth style in the yaml cfg

File: authorization/oauth/oauth.go [hunk 2]
```diff
@@ -30,22 +30,22 @@ func (a AuthStyle) ToOAuth2() oauth2.AuthStyle {
...
+	Audience     string    `yaml:"audience" json:"audience"`
+	RedirectURL  string    `yaml:"redirectURL" json:"redirectURL"`
+	AccessType   string    `yaml:"accessType" json:"accessType"`
+	ResponseType string    `yaml:"responseType" json:"responseType"`
```
Comment (ondrejtomcik) [inline authorization/oauth/oauth.go:44]: this is not supposed to be configurable. shall be skipped from the config

File: authorization/oauth/oauth.go [hunk 3]
```diff
@@ -30,22 +30,22 @@ func (a AuthStyle) ToOAuth2() oauth2.AuthStyle {
...
+type Config struct {
+	ClientID     string    `yaml:"clientID" json:"clientID"`
+	ClientSecret string    `yaml:"clientSecret" json:"clientSecret"`
+	Scopes       []string  `yaml:"scopes" json:"scopes"`
+	AuthURL      string    `yaml:"authorizationURL" json:"authorizationURL"`
+	TokenURL     string    `yaml:"tokenURL" json:"tokenURL"`
+	AuthStyle    AuthStyle `yaml:"authStyle" json:"authStyle"`
+	Audience     string    `yaml:"audience" json:"audience"`
+	RedirectURL  string    `yaml:"redirectURL" json:"redirectURL"`
+	AccessType   string    `yaml:"accessType" json:"accessType"`
```
Comment (ondrejtomcik) [inline authorization/oauth/oauth.go:43]: same here

File: authorization/oauth/oauth.go [hunk 4]
```diff
@@ -30,22 +30,22 @@ func (a AuthStyle) ToOAuth2() oauth2.AuthStyle {
...
+	ResponseType string    `yaml:"responseType" json:"responseType"`
+	ResponseMode string    `yaml:"responseMode" json:"responseMode"`
```
Comment (ondrejtomcik) [inline authorization/oauth/oauth.go:45]: same here

File: authorization/oauth/oauth.go [hunk 5]
```diff
@@ -55,9 +55,9 @@ func (c Config) ToOAuth2() oauth2.Config {
 		RedirectURL:  c.RedirectURL,
 		Scopes:       c.Scopes,
 		Endpoint: oauth2.Endpoint{
-			AuthURL:   c.Endpoint.AuthURL,
-			TokenURL:  c.Endpoint.TokenURL,
-			AuthStyle: c.Endpoint.AuthStyle.ToOAuth2(),
+			AuthURL:   c.AuthURL,
+			TokenURL:  c.TokenURL,
+			AuthStyle: c.AuthStyle.ToOAuth2(),
```
Comment (ondrejtomcik) [inline authorization/oauth/oauth.go:61]: this doesn't exist

File: authorization/persistence/mongodb/config.go [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+type Config struct {
+	URI      string        `yaml:"uri" json:"uri" default:"mongodb://localhost:27017"`
+	Database string        `yaml:"database" json:"database" default:"authorization"`
+	TLS      client.Config `yaml:"tls" json:"tls"`
+}
+
+func (c *Config) Validate() error {
+	if c.URI == "" {
+		return fmt.Errorf("uri")
```
Comment (ondrejtomcik) [inline authorization/persistence/mongodb/config.go]: what uri . be more what is going on?

File: authorization/provider/config.go [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+import (
+	"fmt"
+
+	"github.com/plgd-dev/cloud/authorization/oauth"
+	"github.com/plgd-dev/cloud/pkg/net/http/client"
+)
+
+// Config general configuration
+type Config struct {
+	Provider     string `yaml:"provider" json:"provider" default:"generic"` // value which comes from the device during the sign-up ("apn")
```
Comment (ondrejtomcik) [inline authorization/provider/config.go:12]: is it? what is the property then for configuring providerType?

File: bundle/run.sh [hunk 8]
```diff
@@ -244,10 +255,48 @@ while true; do
 done
     
 # authorization
+## configuration
+### setup root-cas
+while read -r line; do
```
Comment (ondrejtomcik) [inline bundle/run.sh:260]: please create a default yaml file for authorization service and store it to data folder, so it's easy to map a volume, modify it / review it / add your own one.

File: pkg/config/load.go [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+func LoadAndValidateConfig(v Validator) error {
+	err := Load(v)
+	if err != nil {
+		return err
+	}
+	return v.Validate()
+}
+
+// Load loads config from ENV config or arguments config.
```
Comment (ondrejtomcik) [inline pkg/config/load.go:25]: is it loading from env ?

File: authorization/oauth/oauth.go (hunk 1 above)
Comment (jkralik) [reply authorization/oauth/oauth.go:40]: Yes it is hidden. I don't want to change it it now in this MR. We will change it when we move it to coap-gw.

File: authorization/oauth/oauth.go (hunk 2 above)
Comment (jkralik) [reply authorization/oauth/oauth.go:44]: It is hidden in yaml.cfg.  I don't want to change it it now in this MR. We will change it when we move it to coap-gw.

File: authorization/oauth/oauth.go (hunk 3 above)
Comment (jkralik) [reply authorization/oauth/oauth.go:43]: I don't want to change it it now in this MR. We will change it when we move it to coap-gw.

File: authorization/oauth/oauth.go (hunk 4 above)
Comment (jkralik) [reply authorization/oauth/oauth.go:45]: I don't want to change it it now in this MR. We will change it when we move it to coap-gw.

File: authorization/oauth/oauth.go (hunk 5 above)
Comment (jkralik) [reply authorization/oauth/oauth.go:61]: yes it is hardcoded now in auth service

File: authorization/persistence/mongodb/config.go (hunk 6 above)
Comment (jkralik) [reply authorization/persistence/mongodb/config.go]: it's emptry. The finale error message will be like `invalid value: clients.storate.mongoDB.uri`

File: authorization/provider/config.go (hunk 7 above)
Comment (jkralik) [reply authorization/provider/config.go:12]: providerType = provider

File: bundle/run.sh (hunk 8 above)
Comment (jkralik) [reply bundle/run.sh:260]: not possible - we are configure via ENV. we can do it but after we migrate all cfgs.

File: pkg/config/load.go (hunk 9 above)
Comment (jkralik) [reply pkg/config/load.go:25]: Yes you can use env `CONFIG` or arguments `--config` . Both are supported to set path to the cfg file.

File: authorization/oauth/oauth.go (hunk 2 above)
Comment (ondrejtomcik) [reply authorization/oauth/oauth.go:44]: Please add TODO there so it's not forgotten.

File: authorization/persistence/mongodb/config.go (hunk 6 above)
Comment (ondrejtomcik) [reply authorization/persistence/mongodb/config.go]: Important is to see what value is invalid.

File: authorization/oauth/oauth.go (hunk 2 above)
Comment (jkralik) [reply authorization/oauth/oauth.go:44]: Done

File: authorization/persistence/mongodb/config.go (hunk 6 above)
Comment (jkralik) [reply authorization/persistence/mongodb/config.go]: it is only basic validation so only empty values are controlled.

Comment (ondrejtomcik) [reply authorization/persistence/mongodb/config.go]: Then update the message accordingly please.

Comment (jkralik) [reply authorization/persistence/mongodb/config.go]: Done

**Review by ondrejtomcik — APPROVED** [review]

**Comment by ondrejtomcik** [comment]
Don't forget to mention Trishia in squashed merge commit

**Pull request merged**